### PR TITLE
python3Packages.dateparser: run tests

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/pull/115779#issuecomment-795892271

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dotlambda** [comment]
This breaks `python39.pkgs.dateparser`.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115843` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>python39Packages.dateparser: log was empty</li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>python38Packages.dateparser</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
python38Packages.dateparser:
<details>


warning: unused-argument
Unused argument: `flake8`.
Near pkgs/development/python-modules/dateparser/default.nix:5:3:

```
  |
5 | , flake8
  |   ^
```
</details>
</details>

**Comment by dotlambda** [comment]
So, the error boils down to
```
zoneinfo._common.ZoneInfoNotFoundError: 'No time zone found with key US/Eastern'
```
Could it be that the zoneinfo module (which was only introduced in Python 3.9( doesn't have access to time zone data in the build sandbox. I can confirm that, even in `nix-shell --pure -p python39`,
```
$ python3
Python 3.9.2 (default, Feb 19 2021, 13:34:46)
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from zoneinfo import ZoneInfo
>>> ZoneInfo(key='US/Eastern')
zoneinfo.ZoneInfo(key='US/Eastern')
```
is not a problem and that disabling the sandbox makes the tests go through fine.
cc @FRidh @jonringer

**Comment by FRidh** [comment]
`TZPATH` is set in `lib/python3.9/_sysconfigdata__linux_x86_64-linux-gnu.py `.
```
$ nix-shell -p python39 -I nixpkgs=. --pure --run "python3 -c 'import zoneinfo; zoneinfo.ZoneInfo(key=\"US/Eastern\")'"
```
functions as well.
Same as
```
$ nix-shell -p python39.pkgs.setuptools -I nixpkgs=. --pure --run "python3 -c 'import zoneinfo; zoneinfo.ZoneInfo(key=\"US/Eastern\")'"
```
showing `buildPythonPackage` is not causing trouble either.

```
with import ./. {};

runCommand "tztest" {} ''
  ${python39.interpreter} -c "import zoneinfo; zoneinfo.ZoneInfo(key=\"US/Eastern\")"
  touch $out
''
```
succeed as wel for me.

Maybe the tests override `TZPATH`?

**Comment by FRidh** [comment]
Maybe https://github.com/scrapinghub/dateparser/issues/896 ?

**Comment by dotlambda** [comment]
> `TZPATH` is set in `lib/python3.9/_sysconfigdata__linux_x86_64-linux-gnu.py `.

One of the entries for `TZPATH` in that file, and the only one that exists on NixOS, is `/etc/zoneinfo`. That path is not accessible in the sandbox.
Maybe we could prepend `${tzdata}/share/zoneinfo` to `TZPATH` in our build of Python?

**Comment by FRidh** [comment]
But Python already can access `${tzdata}/share/zoneinfo`, even in the sandbox. This must be something in the package or the test. Either case, it needs to be explicitly told where `TZDATA` is. This is in my opinion not sufficiently generic that we should always set it.

**Comment by dotlambda** [comment]
> But Python already can access `${tzdata}/share/zoneinfo`

How? Grepping for `tzdata` in `pkgs/development/interpreters/python` doesn't yield any result.

**Comment by dotlambda** [comment]
According to https://www.python.org/dev/peps/pep-0615/#compile-time-options, we are supposed to set `PYTHONTZPATH` at compile time. The default could be `/etc/zoneinfo:${tzdata}/share/zoneinfo`

**Comment by FRidh** [comment]
On staging, we add `tzdata` as part of the 3.9+ builds. Then, it ends up with a reference in `_sysconfig....` as I showed. The zoneinfo module loads that `TZDATA` variable. Try the Nix snippet I showed, it successfully uses zoneinfo in a build.

**Comment by FRidh** [comment]
PYTHONTZPATH is for additional paths. We don't need additional paths, although maybe we want it to *also* impurely check /etc/zoneinfo. That is an option. But it won't help you with your packages.

**Comment by dotlambda** [comment]
> On staging, we add `tzdata` as part of the 3.9+ builds.

Oh I see. I was looking at master the whole time.
I have to do some rebuilds in order to see if the tests would even fail on staging,

**Comment by FRidh** [comment]
Right I could have been clearer about. Support for tzdata was merged only this weekend.

**Comment by dotlambda** [comment]
It works fine on staging.

File: pkgs/development/python-modules/dateparser/default.nix [hunk 1]
```diff
@@ -1,61 +1,54 @@
...
+  src = fetchFromGitHub {
+    owner = "scrapinghub";
+    repo = "dateparser";
+    rev = "v${version}";
+    sha256 = "0i6ci14lqfsqrmaif57dyilrjbxzmbl98hps1b565gkiy1xqmjhl";
+  };
 
   propagatedBuildInputs = [
     # install_requires
     dateutil pytz regex tzlocal
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dateparser/default.nix:32]: Are they supposed to be not under nativeBuildInputs?

File: pkgs/development/python-modules/dateparser/default.nix [hunk 2]
```diff
@@ -1,61 +1,54 @@
...
     # extra_requires
-    jdatetime ruamel_yaml umalqurra
+    hijri-converter convertdate
   ];
 
+  checkInputs = [
+    parameterized
+    pytestCheckHook
+    GitPython
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dateparser/default.nix:40]: FYI: GitPython is broken with the newest git version with some operations.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115843` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>python38Packages.dateparser</li>
<li>python39Packages.dateparser</li>
</ul>
</details>

**Review by SuperSandro2000 — APPROVED** [review]
other then that LGTM

File: pkgs/development/python-modules/dateparser/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/development/python-modules/dateparser/default.nix:32]: Why would they if they are listed as install_requires?

Comment (SuperSandro2000) [reply pkgs/development/python-modules/dateparser/default.nix:32]: Most python packages don't get there requires right. If it is not required at runtime it should be build or nativeBuildInputs depending if it is executed. If it is required at runtime this is fine.

Comment (dotlambda) [reply pkgs/development/python-modules/dateparser/default.nix:32]: All four are imported outside of the tests.

File: pkgs/development/python-modules/dateparser/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/dateparser/default.nix:40]: No, that part is not used.

File: pkgs/development/python-modules/dateparser/default.nix [hunk 3]
```diff
@@ -1,61 +1,54 @@
...
+    ruamel_yaml
+  ];
+
+  pytestFlagsArray = [ "tests" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dateparser/default.nix:44]:
```suggestion
```
This should be detected automatically.

Comment (dotlambda) [reply pkgs/development/python-modules/dateparser/default.nix:44]: Without it. some tests from `dateparser_scripts/` are executed. I could use `disabledTestPaths` instead.

Comment (dotlambda) [reply pkgs/development/python-modules/dateparser/default.nix:44]: @SuperSandro2000 What do you prefer?

Comment (jonringer) [reply pkgs/development/python-modules/dateparser/default.nix:44]: I think `pytestFlagsArray = [ "tests" ];` is good enough

Comment (SuperSandro2000) [reply pkgs/development/python-modules/dateparser/default.nix:44]: I didn't know some tests from other directories are executed. Max be worth a comment.

Comment (dotlambda) [reply pkgs/development/python-modules/dateparser/default.nix]: done

**Pull request merged**